### PR TITLE
Add CompilerStack state assertions to internal methods

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Review by chriseth — DISMISSED** [review]

**Comment by axic** [comment]
Please double check the actual assertion values.

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -325,6 +325,7 @@ bool CompilerStack::compile()
 
 void CompilerStack::link()
 {
+	solAssert(m_stackState >= AnalysisSuccessful, "");
```
Comment (chriseth) [inline libsolidity/interface/CompilerStack.cpp]: I guess this should be CompilationSuccessful.

**Comment by chriseth** [comment]
I'm sorry, but I don't want to review this any further before 0.5.0 release.

File: libsolidity/interface/CompilerStack.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/interface/CompilerStack.cpp]: True, only based that on what is a required field. It should work after analysis, though it wouldn't contain anything.

**Comment by axic** [comment]
Rebased and fixed the issue noted. I think this should be safe to merge and would prefer merging it now to give us the 2 weeks time to test it.

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
Tests fail.

**Comment by axic** [comment]
Nvm, it seems in the past month something broke:
```
Compiling ./contracts/token/ERC827/ERC827Token.sol...

InternalCompilerError:
Compilation failed. See above.
```

Will investigate.

**Comment by axic** [comment]
It is actually the `link` method, seemingly it is called even when without `compile`.

```
UNCAUGHT EXCEPTION:
- function: "void dev::solidity::CompilerStack::link()"
```

**Comment by axic** [comment]
@chriseth I think it is time to revisit this :)

File: libsolidity/interface/CompilerStack.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/interface/CompilerStack.cpp]: Added.

File: libsolidity/interface/CompilerStack.cpp [hunk 2]
```diff
@@ -329,13 +329,14 @@ bool CompilerStack::compile()
 			if (auto contract = dynamic_cast<ContractDefinition const*>(node.get()))
 				if (isRequestedContract(*contract))
 					compileContract(*contract, compiledContracts);
-	this->link();
 	m_stackState = CompilationSuccessful;
+	this->link();
 	return true;
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp:334]: @chriseth do we want to introduce a `LinkingSuccessful` state?

Comment (chriseth) [reply libsolidity/interface/CompilerStack.cpp:334]: Not sure if that makes sense. An object can have unresolved link symbols and calling `link` does not have to resolve all of them. So in that sense, linking is always successful.

Comment (axic) [reply libsolidity/interface/CompilerStack.cpp:334]: Agree, so lets merge this.

**Review by chriseth — APPROVED** [review]

**Pull request merged**